### ISSUE TITLE
installing from binary cache: relocate_text mangles text

### ISSUE THREAD

**Issue by eugeneswalker** [spack/spack#14262]
@gartung @scottwittenburg 

I've built packages for a binary cache in this source context 
```
SPACK_ROOT=/spack
install_tree=/super/long/obnoxiously/outrageously/absurdly/long/spack/install/path/for/relocation
```

And I'm trying to install these cached binaries into this target context
```
SPACK_ROOT=/tmp/tmp.hI5JOBMHP4/spack
install_tree=/super/absurdly/ridiculously/obnoxiously/long/path/for/relocation
```

But text relocation is not working as expected. For instance, installing a cached binary of autoconf built in the source context into the target context results in the first line of the `autoreconf` script looking like this:
```
#!/bin/bash /tmp/tmp.hI5JOBMHP4/tmp/tmp.hI5JOBMHP4/spack/bin/sbang
```

... which is obviously non-functional.

Tracing the behavior `relocate_text()` from lines 601-611 of `relocate.py`:
```
def relocate_text(path_names, oldpath, newpath, oldprefix, newprefix):
    """
    Replace old path with new path in text files
    including the path the the spack sbang script.
    """
    sbangre = '#!/bin/bash %s/bin/sbang' % oldprefix
    sbangnew = '#!/bin/bash %s/bin/sbang' % newprefix
    for path_name in path_names:
        replace_prefix_text(path_name, oldpath, newpath) # 1 below
        replace_prefix_text(path_name, sbangre, sbangnew) # 2 below
        replace_prefix_text(path_name, oldprefix, newprefix) # 3 below
```

shows the following sequence of substitutions causing the bungled text in the first line of `autoreconf` script:
```
1. 
oldpath = /super/long/obnoxiously/outrageously/absurdly/long/spack/install/path/for/relocation
newpath = /super/absurdly/ridiculously/obnoxiously/long/path/for/relocation
data, pre-sub = #!/bin/bash /spack/bin/sbang\n#! /super/long/obnoxiously/...
data, post-sub = #!/bin/bash /spack/bin/sbang\n#! /super/absurdly/ridiculously...

2.
sbangre = #!/bin/bash /spack/bin/sbang
sbangnew = #!/bin/bash /tmp/tmp.hI5JOBMHP4/spack/bin/sbang
data, pre-sub = #!/bin/bash /spack/bin/sbang ...
data, post-sub = '#!/bin/bash /tmp/tmp.hI5JOBMHP4/spack/bin/sbang ...

3.
oldprefix = /spack
newprefix = /tmp/tmp.hI5JOBMHP4/spack
data, pre-sub = #!/bin/bash /tmp/tmp.hI5JOBMHP4/spack/bin/sbang
data, post-sub = #!/bin/bash /tmp/tmp.hI5JOBMHP4/tmp/tmp.hI5JOBMHP4/spack/bin/sbang
```

I am not very good with regular expressions but perhaps there is a way to adapt `relocate_text()` to  accommodate this type of scenario, where source and destination context are as described above?

**Comment by gartung** [comment]
#3 would be the culprit. This was an attempt to fix the paths for dependencies in text. Using /spack is what really breaks things. The assumption is that spack is no the root of the path. If you used /opt/spack it might work better.

**Comment by gartung** [comment]
You can comment out the third re sub line as a temporary work around. It would be interesting to see if there are any dependency paths remaining.

**Comment by gartung** [comment]
@eugeneswalker are you using a tagged version of spack or the head of develop. It help to know where the fix needs to be applied.

**Comment by eugeneswalker** [comment]
I'm using tag `v0.13.2`

Yes, by the time I tracked down the issue last night my brain was mush. I can try what you suggest, I think it will work. If the answer is just to not use `/spack` that is fine. Although `/spack` is such a sweet final path for installed spack packages, I suppose it's not very important :)

**Comment by gartung** [comment]
Yes. Just remove the third one and see if it works for you.
 I will work on a fix for this is my current branch for buildcache work.

**Comment by gartung** [comment]
The real problem is that the sbang script should be copied and referenced from the install directory root instead of the spack bin directory. I have a  closed PR where I proposed that but it was never merged.
https://github.com/spack/spack/pull/11598

**Comment by gartung** [comment]
Having the sbang script live in what looks like a temporary spack directory that may get removed look like a recipe for trouble.

**Comment by gartung** [comment]
@tgamblin

**Comment by eugeneswalker** [comment]
So @gartung -- I'd like to avoid plugging a custom branch or patch to Spack's code into my pipeline for now: do you think it will work OK if I just change it up like this:
```
SPACK_ROOT=/opt/spack
install_tree=/tmp/blahblah/spack
```

**Comment by gartung** [comment]
As long as blahblah != opt you should be OK.

**Comment by gartung** [comment]
Addressed in #13737

**Comment by gartung** [comment]
https://github.com/spack/spack/pull/13797 should address this
